src/algebra/module/basic.lean

```diff
@@ -461,19 +455,64 @@ end
 
 end add_monoid_hom
 
-section module_division_ring
-/-! Some tests for the vanishing of elements in modules over division rings. -/
+section no_zero_smul_divisors
+/-! ### `no_zero_smul_divisors`
+
+This section defines the `no_zero_smul_divisors` class, and includes some tests
+for the vanishing of elements (especially in modules over division rings).
+-/
+
+/-- `no_zero_smul_divisors R M` states that a scalar multiple is `0` only if either argument is `0`.
+
+The main application of `no_zero_smul_divisors R M`, when `M` is a semimodule,
+is the result `smul_eq_zero`: a scalar multiple is `0` iff either argument is `0`.
+
+It is a generalization of the `no_zero_divisors` class to heterogeneous multiplication.
+-/
+class no_zero_smul_divisors (R M : Type*) [has_zero R] [has_zero M] [has_scalar R M] : Prop :=
+(eq_zero_or_eq_zero_of_smul_eq_zero : ∀ {c : R} {x : M}, c • x = 0 → c = 0 ∨ x = 0)
+
+export no_zero_smul_divisors (eq_zero_or_eq_zero_of_smul_eq_zero)
+
+section semimodule
+
+variables [semiring R] [add_comm_monoid M] [semimodule R M]
+
+instance no_zero_smul_divisors.of_no_zero_divisors [no_zero_divisors R] :
+  no_zero_smul_divisors R R :=
+⟨λ c x, no_zero_divisors.eq_zero_or_eq_zero_of_mul_eq_zero⟩
+
+@[simp]
+theorem smul_eq_zero [no_zero_smul_divisors R M] {c : R} {x : M} :
+  c • x = 0 ↔ c = 0 ∨ x = 0 :=
+⟨eq_zero_or_eq_zero_of_smul_eq_zero,
+ λ h, h.elim (λ h, h.symm ▸ zero_smul R x) (λ h, h.symm ▸ smul_zero c)⟩
+
+section nat
 
-variables (R) [division_ring R] [add_comm_group M] [module R M]
+variables (R) [no_zero_smul_divisors R M] [semimodule ℕ M] [char_zero R]
+include R
 
-lemma smul_nat_eq_zero [semimodule ℕ M] [char_zero R] {v : M} {n : ℕ} :
-  n • v = 0 ↔ n = 0 ∨ v = 0 :=
-by { rw [←nsmul_eq_smul, nsmul_eq_smul_cast R, smul_eq_zero], simp }
+instance nat.no_zero_smul_divisors : no_zero_smul_divisors ℕ M :=
```

Comment:
From what I remember this is a dangerous instance because `R` does not feature in the type